pkg/remover/helpers.go

```diff
@@ -75,6 +76,12 @@ func removeImages(c cri.Remover, targetImages []string) (int, error) {
 				continue
 			}
 
+			// TODO - figure out why is imgDigestOrTag used instead of imageID when it's called "idToImageMap" (copied usage from isExcluded).
+			if !removePinned && util.IsPinned(imageID, idToImageMap) {
```

Comment:
The `util.IsExcluded` function does a `idToImageMap[imgDigestOrTag]`, although I would expect it to use the `imageId` as the key..

Reply:
The reason is a frustrating one: Docker's `ImageID` is different from Containerd's Digest. Newer versions of docker can use containerd to manage storage & metadata of images. 

But without that feature, docker reports the sha256 digest of the Image Config as the `ImageID`. On the other hand, containerd's Image Digest is the sha256 digest of the image manifest. Of the two, the image digest is more stable, because the *name* of the image is included in the Image Config. Thus, a change to the name of an image (without any changes to the image itself) will result in a change to the ImageID.

Thus we use the digest as the hash key for the Set of images we build. Each distinct digest is a distinct image, full stop; the same is not true for ImageIDs: tag an existing image with a new name and you have one distinct image with two distinct ImageIDs.

The CRI is kind of in an in-between state. It was developed to provide an interface and had to work with older clusters using dockershim and newer clusters using containerd. As such, it takes the ImageID into account more than it should.

Using trivy to scan by ImageID doesn't work. Trivy scans the containerd image store by creating a containerd client (from the containerd library) and querying it directly. Since containerd doesn't manage images by docker's ImageID, it can't provide any image information to trivy for the scan if it's looking for it by ImageID.

We want to scan and remove by *content* as much as possible, not by name. We use the image name + tag as a backup in the event that the call to the CRI's `ListImages` returns an image with no digest information. This happens surprisingly frequently because of the ImageID cruft in the CRI implementation.